### PR TITLE
Ambiguous move clarification

### PR DESCRIPTION
**Pull request by jcolemang**
I am using some of the code from Lewis' PR, so this should not be merged before his PR.

Basic ambiguous move clarification. I think it probably should only allow you to specify one of the blocks that was a move candidate, I can add that in the very near future. 

Works nearly identically to Lewis' additions. When an ambiguous move is detected, prompts for a block ID. It then moves/flips that block.

### CODE REVIEW

File: 2D Simulation/game.js [hunk 1]
```diff
@@ -122,10 +122,14 @@ socket.on('enable_blocks_for_player_2', function(data) {
 });
 
 socket.on('update_position', function (moveData) {
+    update_position(moveData);
+});
```
Comment (Lewis-Kelley) [inline 2D Simulation/game.js]: If you're going to pull it out in a function, it's better to use `socket.on('update_position', update_position)` as an eta reduction.

File: 2D Simulation/game.js [hunk 2]
```diff
@@ -154,33 +158,38 @@ function update_gui_block(moveData) {
 }
 
 socket.on('update_flip_block', function (block_id) {
-    let id = block_id.substring(5);
+    update_flip_block(block_id);
+});
```
Comment (Lewis-Kelley) [inline 2D Simulation/game.js]: Again, just call `socket.on('update_flip_block', update_flip_block)` as an eta reduction.

File: 2D Simulation/game.js [hunk 3]
```diff
@@ -154,33 +158,38 @@ function update_gui_block(moveData) {
...
     let color = move['predicted_color'] || 'Ambiguous';
     let letter = move['predicted_letter'] || 'Ambiguous';
     let message = 'I think that this is an ambiguous move. '
-    + 'Please try a more specific instruction.\n'
-    + 'Predicted color: ' + color + '\n'
-    + 'Predicted letter: ' + letter;
+        + 'Please try a more specific instruction.\n'
+        + 'Predicted color: ' + color + '\n'
+        + 'Predicted letter: ' + letter + '\n'
+        + 'Enter the identifier of the intended block.';
```
Comment (Lewis-Kelley) [inline 2D Simulation/game.js]: Should I enter an identifier or a more specific instruction? One of those two statements should go.

File: 2D Simulation/blocks.js [hunk 4]
```diff
@@ -51,5 +51,19 @@ let blocks = new function () {
 
         currentConfig[id].topLetter = newLetter;
         currentConfig[id].bottomLetter = oldLetter;
-    }
+    };
+
+
```
Comment (Lewis-Kelley) [inline 2D Simulation/blocks.js]: Excess whitespace here.

File: 2D Simulation/movesCorrector.js [hunk 5]
```diff
@@ -51,15 +47,9 @@ let movesCorrector = new function () {
...
-        display_block_letters();
+        blocks.display_block_letters();
     };
 
-    function display_block_letters() {
-        for (let i = 0; i < NumBlocks; i++) {
-            blocks.set_block_text(i, currentConfig[i].topLetter);
-        }
-    }
-
```
Comment (Lewis-Kelley) [inline 2D Simulation/movesCorrector.js]: Good call moving these out.

File: 2D Simulation/movesCorrector.js [hunk 6]
```diff
@@ -103,4 +95,32 @@ let movesCorrector = new function () {
     function is_valid_id(id) {
         return id % 1 === 0 && id >= 0 && id < NumBlocks;
     }
+
+    this.is_handling_ambiguity = function() {
+        return handling_ambiguity;
+    };
+
+    this.handle_ambiguity = function(currMove) {
+        console.log('handling ambiguity');
```
Comment (Lewis-Kelley) [inline 2D Simulation/movesCorrector.js]: Do we want this printout here?

File: 2D Simulation/movesCorrector.js [hunk 7]
```diff
@@ -103,4 +95,32 @@ let movesCorrector = new function () {
...
+        handling_ambiguity = true;
+        move = currMove;
+        blocks.display_block_ids();
+    };
+
+    this.finish_handling_ambiguity = function(message) {
+        console.log('this should be doing it.');
```
Comment (Lewis-Kelley) [inline 2D Simulation/movesCorrector.js]: Do we want this printout here?

File: 2D Simulation/movesCorrector.js [hunk 8]
```diff
@@ -103,4 +95,32 @@ let movesCorrector = new function () {
...
+        let id = convert_message_to_id_num(message);
+        move.block_id = 'block' + id;
```
Comment (Lewis-Kelley) [inline 2D Simulation/movesCorrector.js]: We must check if this is a valid `id` before using it.

File: 2D Simulation/movesCorrector.js [hunk 9]
```diff
@@ -103,4 +95,32 @@ let movesCorrector = new function () {
...
+
+        handling_ambiguity = false;
+        blocks.display_block_letters();
+
+        if (move.type == 'ambiguous_move') {
+            move.type = 'move';
+            update_position(move);
+        } else if (move.type == 'ambiguous_flip') {
+            move.type = 'flip';
```
Comment (Lewis-Kelley) [inline 2D Simulation/movesCorrector.js]: This doesn't look like it's used. `update_flip_block` takes `move.block_id`, not `move`.

File: 2D Simulation/game.js [hunk 10]
```diff
@@ -335,6 +345,9 @@ function send_user_message_to_server(gameConfig) {
     if (start_button_pressed) {
         if (movesCorrector.handle_message(message)) {
             return;
+        } else if (movesCorrector.is_handling_ambiguity()) {
+            movesCorrector.finish_handling_ambiguity(message);
```
Comment (Lewis-Kelley) [inline 2D Simulation/game.js]: I feel like this should go inside the `handle_message` method. Let the `movesCorrector` figure out which state it's in, not the calling code.

**Review by Lewis-Kelley — CHANGES_REQUESTED** [review]
Most of the comments are cosmetic. The only super-important one is that you don't validate the `id` before using it in `finish_handling_ambiguity`.

File: 2D Simulation/movesCorrector.js (hunk 9 above)
Comment (jcolemang) [reply 2D Simulation/movesCorrector.js]: It isn't, but having a move type equal to `"ambiguous_flip"` when it is no longer ambiguous seems bad to me.

File: 2D Simulation/game.js (hunk 10 above)
Comment (jcolemang) [reply 2D Simulation/game.js]: I disagree. Calling `finish_handling_ambiguity` seems bad if it isn't going to handle any ambiguity. That seems like weird control flow to me.

**Comment by jcolemang** [comment]
I messed around with the base branch with involved some merge conflicts. Keep that in mind when anyone is reviewing this

File: 2D Simulation/game.js (hunk 10 above)
Comment (Lewis-Kelley) [reply 2D Simulation/game.js]: I'm more talking about moving:
```
if (movesCorrector.is_handling_ambiguity()) {
    movesCorrector.finish_handling_ambiguity(message);
}
```
 into the `handle_message` method. Everything else would stay the same; `finish_handling_ambiguity` could then be a private method.

**Review by Lewis-Kelley — APPROVED** [review]

File: 2D Simulation/game.js (hunk 10 above)
Comment (jcolemang) [reply 2D Simulation/game.js]: I guess I misunderstood the purpose of that method. I can move it in there.

**Comment by jcolemang** [comment]
There were more conflicts from the refactoring commits that were merged in. I added a commit to merge them in but still need to check that I didn't break anything.

**Review by kgroble — APPROVED** [review]

**Pull request merged**